### PR TITLE
[TT-7815] Migrate API with endpoints containing path parameter 

### PR DESCRIPTION
**Pull request by edsonmichaque**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-7815" title="TT-7815" target="_blank">TT-7815</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Cannot migrate API with endpoints containing path parameter </td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td><a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20DoD_Fail%20ORDER%20BY%20created%20DESC" title="DoD_Fail">DoD_Fail</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20Re_open%20ORDER%20BY%20created%20DESC" title="Re_open">Re_open</a></td>
    </tr>
  </table>
</details>

---

## Related Issue

https://tyktech.atlassian.net/browse/TT-7815

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Bug fix, Tests

___

### **Description**
- Enhanced path parameter handling in API definitions.

- Added support for named and regex-based path parameters.

- Refactored `splitPath` logic for improved clarity and functionality.

- Introduced comprehensive unit tests for path parsing and validation.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>operation.go</strong><dd><code>Refactor path parsing and parameter handling logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/operation.go

<li>Added <code>isParam</code> field to <code>pathPart</code> struct.<br> <li> Enhanced <code>splitPath</code> to handle named and regex-based parameters.<br> <li> Refactored <code>isRegex</code> and introduced <code>isParamName</code> for validation.<br> <li> Improved <code>buildPath</code> to support new parameter types.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6901/files#diff-6d92d2d5b09a5fa7129609bb7cd0d383d015250ec07062b6a93a83257be51fb5">+104/-20</a></td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>operation_test.go</strong><dd><code>Add unit tests for path parsing and validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/operation_test.go

<li>Added new test cases for <code>splitPath</code> and <code>pathPart.String</code>.<br> <li> Enhanced <code>TestOAS_RegexPaths</code> to validate parameter extraction.<br> <li> Introduced tests for complex path scenarios and edge cases.<br> <li> Improved test coverage for named and regex-based parameters.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6901/files#diff-cd234db716d6d2edc97c135ef546021c9ab4fa9282d63964bd155d41635cf964">+295/-20</a></td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by buger** [comment]
<p>
    Let's make that PR title a 💯 shall we? 💪
    </p>
    <p>
    Your <em>PR title</em> and <em>story title</em> look <strong>slightly different</strong>. Just checking in to know if it was intentional!
    </p>
    <table>
      <tr>
        <th>Story Title</th>
        <td>Cannot migrate API with endpoints containing path parameter </td>
      </tr>
      <tr>
          <th>PR Title</th>
          <td>[TT-7815] migrate path params</td>
        </tr>
    </table>
    <p>
      Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
    </p>

File: apidef/oas/operation.go [hunk 1]
```diff
@@ -501,27 +502,115 @@ func isRegex(value string) bool {
...
+
+				result[k].name = fmt.Sprintf("customRegex%d", nCustomRegex)
+				result[k].isRegex = true
+			}
+
+			continue
+		}
+
+		// Handle bracketed path segments
+		segment := trimPathParam(value)
```
Comment (titpetric) [inline apidef/oas/operation.go:551]:
```suggestion
		segment := strings.Trim(value, "{}")
```

File: apidef/oas/operation.go [hunk 2]
```diff
@@ -501,27 +502,115 @@ func isRegex(value string) bool {
...
+		return value
+	}
 
 	for k, value := range parts {
-		name := value
-		isRegex := isRegex(value)
-		if isRegex {
+		// Handle non-bracketed path segments
+		// for example: /a/b/c, /a/[0-9]
+		if !strings.HasPrefix(value, "{") && !strings.HasSuffix(value, "}") {
```
Comment (titpetric) [inline apidef/oas/operation.go:531]: httputil package has better implementation for this step (strings.Count...)

File: apidef/oas/operation.go [hunk 3]
```diff
@@ -866,3 +960,10 @@ func (s *OAS) fillCircuitBreaker(metas []apidef.CircuitBreakerMeta) {
 		}
 	}
 }
+
+// validatePath validates if the path is valid. Returns an error.
+// TODO: This is a temporary implementation to avoid circular dependency.
+// Should be refactored to use httputil.ValidatePath once dependency structure is resolved.
+func validatePath(in string) error {
+	return mux.NewRouter().PathPrefix(in).GetError()
```
Comment (titpetric) [inline apidef/oas/operation.go:968]: You get the regex with .Getregexppath, this is not enough right?

You want to use mux to convert `/{id}` to a named matching group that matches mux.
isParamName should go away as this would report an error in the pattern.

File: apidef/oas/operation.go [hunk 4]
```diff
@@ -501,27 +502,115 @@ func isRegex(value string) bool {
...
+	if len(s) == 0 {
+		return false
+	}
+
+	// Check if string ends with hyphen
+	if s[len(s)-1] == '-' {
+		return false
+	}
+
+	for i, c := range s {
```
[the remark this answers is not in the document]
Comment (edsonmichaque) [reply apidef/oas/operation.go:603]: Fixed it, thanks. PTAL

File: apidef/oas/operation.go (hunk 2 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed it, thanks

File: apidef/oas/operation.go (hunk 1 above)
Comment (titpetric) [reply apidef/oas/operation.go:685]: Yeah, I don't think anything else is valid, because it would be `{{id}}` or `{{id}` or `{id:...}}` (all invalid).

File: apidef/oas/operation.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (edsonmichaque) [reply apidef/oas/operation.go]: I've removed that function, we don't need it anymore

File: apidef/oas/operation.go (hunk 1 above)
Comment (edsonmichaque) [reply apidef/oas/operation.go]: Fixed, PTAL

File: apidef/oas/operation.go [hunk 5]
```diff
@@ -620,51 +622,116 @@ type pathPart struct {
...
+
+		// Handle bracketed path segments
+		segment := strings.Trim(value, "{}")
+
+		// Parameter with pattern case:
+		// for example: /a/{id:[0-9]}
+		if name, pattern, ok := strings.Cut(segment, ":"); ok && strings.TrimSpace(name) != "" {
+			// this is required because GetPathRegexp doesn't check if the regex is valid
+			// it doesn't return an error even if the regex is invalid (eg. [a-zA-Z+)
+			if _, err := regexp.Compile(pattern); err != nil {
```
Comment (titpetric) [inline apidef/oas/operation.go:690]: change the import to this to the /regexp package, caching the compilation

Comment (edsonmichaque) [reply apidef/oas/operation.go:691]: Done

File: apidef/oas/operation_test.go [hunk 6]
```diff
@@ -1113,3 +1165,295 @@ func verifyOASOperation(t *testing.T, op *openapi3.Operation, method string, cod
...
+			expectedParts: []pathPart{
+				{name: "users", value: "users", isRegex: false},
+				{name: "id", value: "[0-9]+", isRegex: true},
+				{name: "profile", value: "profile", isRegex: false},
+			},
+			expectedRegex: true,
+		},
+		{
+			name: "path with direct regex",
+			path: "/files/{.*}/download",
```
Comment (titpetric) [inline apidef/oas/operation_test.go:1262]: this is invalid by mux; read through httputil/mux* ; i feel a significant part is already implemented in httputil, maybe needs extending (mux.go, we already have logic for `/*` and `/*/` which equals this patch matching case.

With `{.*}` - this mux tag is not a valid input, while `{_}` or `{_:.*}` would be. Is this a case we need to cover? Technically it would "work" in mux because `.*` would be treated as a `name`, using the default parameter regex. customRegex in this case should ignore {} and be isRegex=false

File: apidef/oas/operation_test.go [hunk 7]
```diff
@@ -1113,3 +1165,295 @@ func verifyOASOperation(t *testing.T, op *openapi3.Operation, method string, cod
...
+			wantErr: false,
+		},
+		{
+			name:    "valid path with regex parameter",
+			path:    "/api/users/{id:[0-9]+}",
+			wantErr: false,
+		},
+		{
+			name:    "invalid path - unclosed brace",
+			path:    "/api/users/{id",
```
Comment (titpetric) [inline apidef/oas/operation_test.go:1426]: Rather you extend some existing tests for httputil.mux with these; if you can, patch the logic there

File: apidef/oas/operation_test.go (hunk 6 above)
Comment (jeffy-mathew) [reply apidef/oas/operation_test.go:1262]: @titpetric qq: [IsMuxTemplate](https://github.com/TykTechnologies/tyk/blob/83d4b12c7fb8bb7ea3a1caf2489cff4f4722a119/internal/httputil/mux.go#L73) is checking only for existence of curly braces which are opened and closed. And if it is returning true, whatever is inside `{}` is replaced with `([^/]+)`. 
This means `{id:[a-z]}` is also replaced with `([^/]+)`, is it expected?

**Comment by edsonmichaque** [comment]
After refinement of https://tyktech.atlassian.net/browse/TT-7815 I am closing this PR and replacing it by https://github.com/TykTechnologies/tyk/pull/6966
